LobbyServer2/LobbyServer/Matchmaking/MatchmakingQueue.cs

```diff
@@ -11,59 +14,108 @@ namespace CentralServer.LobbyServer.Matchmaking
     class MatchmakingQueue
     {
         Dictionary<string, LobbyGameInfo> Games = new Dictionary<string, LobbyGameInfo>();
-        SynchronizedCollection<LobbyServerProtocolBase> QueuedPlayers = new SynchronizedCollection<LobbyServerProtocolBase>();
-        GameType GameType;
+        SynchronizedCollection<LobbyServerProtocolBase> Players = new SynchronizedCollection<LobbyServerProtocolBase>();
+        public LobbyMatchmakingQueueInfo MatchmakingQueueInfo;
+        GameType GameType => MatchmakingQueueInfo.GameType;
         
         private static int GameID = 0;
 
         public MatchmakingQueue(GameType gameType)
         {
-            GameType = gameType;
+            MatchmakingQueueInfo = new LobbyMatchmakingQueueInfo()
+            {
+                QueueStatus = QueueStatus.Idle,
+                QueuedPlayers = 0,
+                ShowQueueSize = true,
+                AverageWaitTime = TimeSpan.FromSeconds(60),
+                GameConfig = new LobbyGameConfig()
+                {
+                    GameType = gameType,
+                    SubTypes = GameModeManager.GetGameTypeAvailabilities()[gameType].SubTypes
+                }
+            };
+        }
+
+        public LobbyMatchmakingQueueInfo AddPlayer(LobbyServerProtocolBase connection)
+        {
+            Players.Add(connection);
+            MatchmakingQueueInfo.QueuedPlayers = Players.Count;
+
+            return MatchmakingQueueInfo;
         }
 
         public void Update()
         {
-			for (int i = 0; i < QueuedPlayers.Count; i++)
+            RemoveDisconnectedPlayers();
+            Log.Print(LogType.Game, $"{Players.Count} players in {GameType} queue");
+
+            foreach (GameSubType subType in MatchmakingQueueInfo.GameConfig.SubTypes)
+            {
+                int FullGamePlayers = subType.TeamAPlayers + subType.TeamBPlayers;
+                if (Players.Count >= FullGamePlayers)
+                {
+                    if (!CheckGameServerAvailable())
```

Comment:
The whole update should happen under a lock to avoid issues with multiple games statring at once, though I guess we are not quite there yet. 
Also it's kinda weird that function named Check launches servers.